src/utilities/sortIntrospection.js

```diff
@@ -0,0 +1,123 @@
+ *
+ * @flow
+ */
+
+import type {
+  IntrospectionQuery,
+  IntrospectionType,
+  IntrospectionField,
+  IntrospectionDirective,
+} from './introspectionQuery';
+
+/**
+ * Sort the result of Introspection Query.
+ */
+export function sortIntrospectionQuery(
+  introspection: IntrospectionQuery,
+): IntrospectionQuery {
+  const schema = introspection.__schema;
+  return {
+    __schema: {
+      queryType: schema.queryType,
+      mutationType: schema.mutationType,
+      subscriptionType: schema.subscriptionType,
+      types: sortByName(schema.types).map(sortType),
+      directives: sortByName(schema.directives).map(sortDirective),
+    },
+  };
+}
+
+function sortDirective(
+  directive: IntrospectionDirective,
+): IntrospectionDirective {
+  return {
+    name: directive.name,
+    description: directive.description,
+    locations: sortBy(directive.locations, x => x),
+    args: sortByName(directive.args),
+  };
+}
+
+function sortType(type: IntrospectionType): IntrospectionType {
+  switch (type.kind) {
+    case 'OBJECT':
+      return {
+        kind: 'OBJECT',
+        name: type.name,
+        description: type.description,
+        fields: sortFields(type.fields),
+        interfaces: sortByName(type.interfaces),
+      };
+    case 'INTERFACE':
+      return {
+        kind: 'INTERFACE',
+        name: type.name,
+        description: type.description,
+        fields: sortFields(type.fields),
+        possibleTypes: sortByName(type.possibleTypes),
+      };
+    case 'UNION':
+      return {
+        kind: 'UNION',
+        name: type.name,
+        description: type.description,
+        possibleTypes: sortByName(type.possibleTypes),
+      };
+    case 'ENUM':
+      return {
+        kind: 'ENUM',
+        name: type.name,
+        description: type.description,
+        enumValues: sortByName(type.enumValues),
+      };
+    case 'INPUT_OBJECT':
+      return {
+        kind: 'INPUT_OBJECT',
+        name: type.name,
+        description: type.description,
+        inputFields: sortByName(type.inputFields),
+      };
+    case 'SCALAR':
+      return type;
+    default:
+      throw new Error('Unexpected value kind: ' + (type.kind: empty));
+  }
+}
+
+function sortFields(
+  fields: $ReadOnlyArray<IntrospectionField>,
+): $ReadOnlyArray<IntrospectionField> {
+  return sortByName(fields).map(field => ({
+    name: field.name,
+    description: field.description,
+    args: sortByName(field.args),
+    type: field.type,
+    isDeprecated: field.isDeprecated,
+    deprecationReason: field.deprecationReason,
+  }));
+}
+
+function sortByName<T: { +name: string }>(
+  array: $ReadOnlyArray<T>,
+): $ReadOnlyArray<T> {
+  return sortBy(array, obj => obj.name);
+}
+
+function sortBy<T>(
+  array: $ReadOnlyArray<T>,
+  mapToKey: T => string,
+): $ReadOnlyArray<T> {
+  if (!Array.isArray(array)) {
+    throw new Error('Expected array got: ' + (array: empty));
+  }
+  return array.slice().sort((obj1, obj2) => {
```

Comment:
a bit of a nit, but we can remove this `slice` with a bit of trickery. We also will no longer need the special `sortBy<T>`.

```
function sortFields(
  fields: $ReadOnlyArray<IntrospectionField>,
): $ReadOnlyArray<IntrospectionField> {
  return sortByName(fields.map(field => ({
    name: field.name,
    description: field.description,
    args: sortByName(field.args.slice()),
    type: field.type,
    isDeprecated: field.isDeprecated,
    deprecationReason: field.deprecationReason,
  }));
 }

// This cannot be called on an original ReadOnly array, or it would mutate them
function sortByName<T: { +name: string }>(
  array: Array<T>,
): $ReadOnlyArray<T> {
  return array.sort((obj1, obj2) => obj1.name.localeCompare(obj2.name));
}
```

This ends up being a little more code (you now have to `slice()` whenever you want to `sortByName`), but now we save one array copy per field list.